### ISSUE TITLE
Should discovery be reframed as a selectable set of discovery strategies?

### ISSUE THREAD

**Issue by cdaylward** [appc/spec#140]
Should the discovery process be as prescriptive as it is? I find the discovery mechanisms currently outlined in the specification to be well thought out, however I would like to see discovery reframed as "discovery strategies". Simple discovery is a strategy. Meta-discovery is a strategy. DNS discovery could be a strategy. I believe it would be beneficial to outline the structure of the various strategies, but not prescribe which ones must be used or in which order (or perhaps only a partial order).

As the specification stands now, I imagine some implementations will likely drop discovery as it is currently outlined. For instance, [simple discovery](https://github.com/appc/spec/blob/master/SPEC.md#simple-discovery) might simply be disallowed (production networks might forbid "going out to the internet"). [Meta-discovery](https://github.com/appc/spec/blob/master/SPEC.md#meta-discovery) might also not be necessary and a simple local static lookup table might be preferred (e.g. using configuration files or known directory structure for no-network discovery). Additional proposals such as the adoption of something like DNS discovery as discussed in issue/#3 might also be inappropriate. Users who want to use appc tech to share images on the internet might simply not have access to such infrastructure. These things might lead to either lower adoption or simply ignoring this part of the specification.

If the specification limited itself to outlining common strategies and how they should be structured (e.g. if labels are required in lookup, how they are used in lookup (e.g. key is of form `{name}-{version}-{os}-{arch}`), dependency matching, etc), this might promote the development and sharing of different discovery schemes/libraries/tools.

**Comment by jonboulle** [comment]
+1, as the spec has evolved I would argue that the utility of simple discovery has diminished (and conversely, meta discovery has become more powerful), so I would definitely like to decouple those two for starters.

**Comment by cnelson** [comment]
+1

**Comment by philips** [comment]
I agree with the premise and think we should describe the label matching idea in general, then different mechanisms. I also think it would be cool to have an offline crawler/cache system like what exists at godoc.org so that you would tell the discovery mechanism: "prefix everything from aci-index.corp/" before doing the discovery process.

**Comment by philips** [comment]
@cdaylward Do you want to take a swing at describing a file/table based method, the idea of a "behind firewall" prefixing system and separate out the generic idea of the template matching from the generic concept of the discovery mechanism?

**Comment by sgotti** [comment]
+1.

A note: just only with simple and meta discovery some doubts exited due to discrepancies betwenn these two methods. For example:

* How to implement possible use cases with discovery + label matching? See #117.
* Should the above use cases be the base for the definition of all the discovery strategies? (or it will end with a strategy working in a different way than another)

* What urls should be provided by the discovery strategies? (ACI URls, Sig URls and PubKey URL?) (for example simple discovery doesn't support public keys discovery) (I tried to a clarify it a little in the spec in #137)

> I agree with the premise and think we should describe the label matching idea in general

I also think that the label matching idea should be described with various use cases in mind. One of these is the "latest pattern".
If you are interested I already tried to explain some doubts about it here: https://github.com/appc/spec/issues/60#issuecomment-69899696 (and also #73 as the latest pattern is used also for dependency matching)

**Comment by cdaylward** [comment]
@philips Sure, I will try to take a swing at it soon. Lot's of things in motion currently...

**Comment by philips** [comment]
@cdaylward ping :)

**Comment by jonboulle** [comment]
@cdaylward think you'll have time for this in O(1-2weeks)?

**Comment by cdaylward** [comment]
Yes. I'll be able to pick this back up.

**Comment by jonboulle** [comment]
@cdaylward Check out the discussion in https://github.com/coreos/rkt/issues/695 too

**Comment by jonboulle** [comment]
bumping to 0.7.0, @cdaylward is going to look at this reeeeall soon now :-)
